lib/std/hash/xxhash.zig

```diff
@@ -46,99 +142,95 @@ pub const XxHash64 = struct {
-        self.acc4 = round(self.acc4, mem.readIntLittle(u64, buf[24..32]));
-        self.byte_count += 32;
-    }
-
     inline fn round(acc: u64, lane: u64) u64 {
         const a = acc +% (lane *% prime_2);
         const b = rotl(u64, a, 31);
         return b *% prime_1;
     }
 
     pub fn final(self: *XxHash64) u64 {
-        var acc: u64 = undefined;
+        const unfinished = if (self.byte_count < 32)
+            self.seed +% prime_5
+        else
+            self.accumulator.merge();
 
-        if (self.byte_count < 32) {
-            acc = self.seed +% prime_5;
-        } else {
-            acc = rotl(u64, self.acc1, 1) +% rotl(u64, self.acc2, 7) +%
-                rotl(u64, self.acc3, 12) +% rotl(u64, self.acc4, 18);
-            acc = mergeAccumulator(acc, self.acc1);
-            acc = mergeAccumulator(acc, self.acc2);
-            acc = mergeAccumulator(acc, self.acc3);
-            acc = mergeAccumulator(acc, self.acc4);
-        }
+        return finalize(unfinished, self.byte_count, self.buf[0..self.buf_len]);
+    }
 
-        acc = acc +% @as(u64, self.byte_count) +% @as(u64, self.buf_len);
+    const Size = enum {
+        small,
+        large,
+        unknown,
+    };
 
-        var pos: usize = 0;
-        while (pos + 8 <= self.buf_len) : (pos += 8) {
-            const lane = mem.readIntLittle(u64, self.buf[pos..][0..8]);
-            acc ^= round(0, lane);
-            acc = rotl(u64, acc, 27) *% prime_1;
-            acc +%= prime_4;
-        }
-
-        if (pos + 4 <= self.buf_len) {
-            const lane = @as(u64, mem.readIntLittle(u32, self.buf[pos..][0..4]));
-            acc ^= lane *% prime_1;
-            acc = rotl(u64, acc, 23) *% prime_2;
-            acc +%= prime_3;
-            pos += 4;
-        }
+    pub fn hash(seed: u64, input: anytype) u64 {
+        validateType(@TypeOf(input));
 
-        while (pos < self.buf_len) : (pos += 1) {
-            const lane = @as(u64, self.buf[pos]);
-            acc ^= lane *% prime_5;
-            acc = rotl(u64, acc, 11) *% prime_1;
+        if (input.len < 32) {
+            return finalize(seed +% prime_5, 0, input);
+        } else {
+            var hasher = Accumulator.init(seed);
+            const i = hasher.updateEmpty(input, 0);
+            return finalize(hasher.merge(), i, input[i..]);
         }
-
-        acc ^= acc >> 33;
-        acc *%= prime_2;
-        acc ^= acc >> 29;
-        acc *%= prime_3;
-        acc ^= acc >> 32;
-
-        return acc;
-    }
-
-    inline fn mergeAccumulator(acc: u64, other: u64) u64 {
-        const a = acc ^ round(0, other);
-        const b = a *% prime_1;
-        return b +% prime_4;
     }
 
-    pub fn hash(seed: u64, input: []const u8) u64 {
-        var hasher = XxHash64.init(seed);
-        hasher.update(input);
-        return hasher.final();
+    pub const small_key_max_size = 31;
+
+    pub fn hashSmall(seed: u64, input: []const u8) u64 {
```

Comment:
As you've commented already but this feels a little clunky. I prefer the `anytype` approach and relying on the user to pass the specific array with comptime-known length if they require that; assuming it has equal performance.

Would be interested in the codesize difference in doing the below vs. the runtime gain.

Reply:
At the moment I mostly have the `hashSmall` function there as a point of reference for how fast you can hash small keys when you know size is small (letting you avoid a branch), but don't know the size statically. It is however showing some odd results in the benchmark - `hashSmall()` on a `[]u8` of length 16 apparently hash higher throughput than `hash()` on a `[16]u8` (by >30%) or `*[16]u8` (~5-7%). This seems to either reveal some weakness of the `anytype` approach, or a problem in optimisations llvm is doing.

Reply:
I  believe I closed the performance difference between  `hashSmall` and `hash([N]u8)` and have subsequently removed the `hashSmall` API.